File: code/fsharp/grpc/server/TaskService.fs
module TaskGrpcServer.TaskService

open System
open System.Collections.Concurrent
open System.Linq
open Microsoft.Extensions.Logging
open Grpc.Core
open Google.Protobuf.WellKnownTypes
open TaskNamespace

type TaskServiceImplementation(logger: ILogger<TaskServiceImplementation>) =
    inherit TaskService.TaskServiceBase()
    
    let tasks = ConcurrentDictionary<string, Task>()
    
    // Initialize with sample data using F# immutable approach
    let initializeSampleData () =
        let now = Timestamp.FromDateTime(DateTime.UtcNow)
        
        let sampleTasks = [
            Task(
                Id = Guid.NewGuid().ToString(),
                Title = "Complete F# project documentation",
                Description = "Write comprehensive documentation for the F# gRPC API",
                Status = TaskStatus.InProgress,
                Priority = TaskPriority.High,
                AssignedTo = "dev-team",
                CreatedBy = "system",
                CreatedAt = now,
                UpdatedAt = now
            ) |> fun task -> 
                task.Tags.AddRange(["documentation"; "fsharp"; "api"])
                task
            
            Task(
                Id = Guid.NewGuid().ToString(),
                Title = "Review functional programming patterns",
                Description = "Review and approve functional programming patterns in codebase",
                Status = TaskStatus.Pending,
                Priority = TaskPriority.Medium,
                AssignedTo = "senior-dev",
                CreatedBy = "system",
                CreatedAt = now,
                UpdatedAt = now
            ) |> fun task ->
                task.Tags.AddRange(["review"; "functional"; "patterns"])
                task
                
            Task(
                Id = Guid.NewGuid().ToString(),
                Title = "Deploy F# services to production",
                Description = "Deploy the F# microservices to production environment",
                Status = TaskStatus.Pending,
                Priority = TaskPriority.Critical,
                AssignedTo = "devops",
                CreatedBy = "system",
                CreatedAt = now,
                UpdatedAt = now
            ) |> fun task ->
                task.Tags.AddRange(["deployment"; "production"; "fsharp"])
                task
        ]
        
        sampleTasks 
        |> List.iter (fun task -> tasks.[task.Id] <- task)
    
    // Initialize sample data in constructor
    do initializeSampleData()
    
    // Helper functions using F# functional programming
    let validateTaskId (id: string) =
        match String.IsNullOrEmpty(id) with
        | true -> Error (RpcException(Status(StatusCode.InvalidArgument, "Task ID is required")))
        | false -> Ok id
    
    let validateTaskTitle (title: string) =
        match String.IsNullOrWhiteSpace(title) with
        | true -> Error (RpcException(Status(StatusCode.InvalidArgument, "Title is required")))
        | false when title.Length > 200 -> Error (RpcException(Status(StatusCode.InvalidArgument, "Title must be 200 characters or less")))
        | false -> Ok title
    
    let findTask (id: string) =
        match tasks.TryGetValue(id) with
        | true, task -> Ok task
        | false, _ -> Error (RpcException(Status(StatusCode.NotFound, sprintf "Task with ID '%s' not found" id)))
    
    let applyFilters (request: ListTasksRequest) (taskSeq: Task seq) =
        taskSeq
        |> fun tasks -> 
            if request.Status <> TaskStatus.Unspecified then
                tasks |> Seq.filter (fun t -> t.Status = request.Status)
            else tasks
        |> fun tasks ->
            if not (String.IsNullOrEmpty(request.AssignedTo)) then
                tasks |> Seq.filter (fun t -> t.AssignedTo = request.AssignedTo)
            else tasks
        |> fun tasks ->
            if request.Tags.Count > 0 then
                tasks |> Seq.filter (fun t -> 
                    request.Tags |> Seq.exists (fun tag -> t.Tags.Contains(tag)))
            else tasks
    
    let applySorting (request: ListTasksRequest) (taskSeq: Task seq) =
        match request.SortOrder with
        | SortOrder.CreatedAsc -> taskSeq |> Seq.sortBy (fun t -> t.CreatedAt.ToDateTime())
        | SortOrder.CreatedDesc -> taskSeq |> Seq.sortByDescending (fun t -> t.CreatedAt.ToDateTime())
        | SortOrder.UpdatedAsc -> taskSeq |> Seq.sortBy (fun t -> t.UpdatedAt.ToDateTime())
        | SortOrder.UpdatedDesc -> taskSeq |> Seq.sortByDescending (fun t -> t.UpdatedAt.ToDateTime())
        | SortOrder.PriorityDesc -> 
            taskSeq 
            |> Seq.sortByDescending (fun t -> int t.Priority)
            |> Seq.thenBy (fun t -> t.CreatedAt.ToDateTime())
        | _ -> taskSeq |> Seq.sortByDescending (fun t -> t.CreatedAt.ToDateTime())
    
    let applyPagination (request: ListTasksRequest) (taskSeq: Task seq) =
        let pageSize = if request.PageSize > 0 then min (int request.PageSize) 100 else 20
        let skip = 
            match String.IsNullOrEmpty(request.PageToken) with
            | true -> 0
            | false -> 
                match Int32.TryParse(request.PageToken) with
                | true, value -> value
                | false, _ -> 0
        
        taskSeq |> Seq.skip skip |> Seq.take pageSize
    
    override this.GetTask(request: GetTaskRequest, context: ServerCallContext) =
        async {
            logger.LogInformation("GetTask called with ID: {TaskId}", request.Id)
            
            return
                result {
                    let! _ = validateTaskId request.Id
                    let! task = findTask request.Id
                    return task
                }
                |> function
                | Ok task -> task
                | Error ex -> raise ex
        } |> Async.StartAsTask
    
    override this.ListTasks(request: ListTasksRequest, responseStream: IServerStreamWriter<Task>, context: ServerCallContext) =
        async {
            logger.LogInformation("ListTasks called")
            
            try
                let filteredTasks = 
                    tasks.Values
                    |> applyFilters request
                    |> applySorting request
                    |> applyPagination request
                
                for task in filteredTasks do
                    if context.CancellationToken.IsCancellationRequested then
                        break
                    do! responseStream.WriteAsync(task) |> Async.AwaitTask
                    
            with
            | ex -> 
                logger.LogError(ex, "Error in ListTasks")
                raise (RpcException(Status(StatusCode.Internal, "Internal server error")))
        } |> Async.StartAsTask :> System.Threading.Tasks.Task
    
    override this.CreateTask(request: CreateTaskRequest, context: ServerCallContext) =
        async {
            logger.LogInformation("CreateTask called")
            
            return
                result {
                    do! 
                        match request.Task with
                        | null -> Error (RpcException(Status(StatusCode.InvalidArgument, "Task is required")))
                        | _ -> Ok ()
                    
                    let! _ = validateTaskTitle request.Task.Title
                    
                    let now = Timestamp.FromDateTime(DateTime.UtcNow)
                    let task = Task(
                        Id = Guid.NewGuid().ToString(),
                        Title = request.Task.Title,
                        Description = (if String.IsNullOrEmpty(request.Task.Description) then "" else request.Task.Description),
                        Status = TaskStatus.Pending,
                        Priority = (if request.Task.Priority = TaskPriority.Unspecified then TaskPriority.Medium else request.Task.Priority),
                        AssignedTo = (if String.IsNullOrEmpty(request.Task.AssignedTo) then "" else request.Task.AssignedTo),
                        CreatedBy = (if String.IsNullOrEmpty(request.Task.CreatedBy) then "system" else request.Task.CreatedBy),
                        CreatedAt = now,
                        UpdatedAt = now,
                        DueDate = request.Task.DueDate
                    )
                    
                    task.Tags.AddRange(request.Task.Tags)
                    tasks.[task.Id] <- task
                    
                    logger.LogInformation("Created task {TaskId}: {Title}", task.Id, task.Title)
                    return task
                }
                |> function
                | Ok task -> task
                | Error ex -> raise ex
        } |> Async.StartAsTask
    
    override this.UpdateTask(request: UpdateTaskRequest, context: ServerCallContext) =
        async {
            logger.LogInformation("UpdateTask called with ID: {TaskId}", 
                (if request.Task <> null then request.Task.Id else "null"))
            
            return
                result {
                    do!
                        match request.Task with
                        | null -> Error (RpcException(Status(StatusCode.InvalidArgument, "Task with ID is required")))
                        | task when String.IsNullOrEmpty(task.Id) -> Error (RpcException(Status(StatusCode.InvalidArgument, "Task ID is required")))
                        | _ -> Ok ()
                    
                    let! existingTask = findTask request.Task.Id
                    
                    // Create updated task using functional approach
                    let updatedTask = Task()
                    updatedTask.Id <- existingTask.Id
                    updatedTask.Title <- existingTask.Title
                    updatedTask.Description <- existingTask.Description
                    updatedTask.Status <- existingTask.Status
                    updatedTask.Priority <- existingTask.Priority
                    updatedTask.AssignedTo <- existingTask.AssignedTo
                    updatedTask.CreatedBy <- existingTask.CreatedBy
                    updatedTask.CreatedAt <- existingTask.CreatedAt
                    updatedTask.UpdatedAt <- Timestamp.FromDateTime(DateTime.UtcNow)
                    updatedTask.DueDate <- existingTask.DueDate
                    updatedTask.Tags.AddRange(existingTask.Tags)
                    
                    // Apply updates based on update mask using pattern matching
                    request.UpdateMask
                    |> Seq.iter (fun field ->
                        match field with
                        | "title" -> 
                            match validateTaskTitle request.Task.Title with
                            | Ok _ -> updatedTask.Title <- request.Task.Title
                            | Error ex -> raise ex
                        | "description" -> updatedTask.Description <- request.Task.Description
                        | "status" -> updatedTask.Status <- request.Task.Status
                        | "priority" -> updatedTask.Priority <- request.Task.Priority
                        | "tags" -> 
                            updatedTask.Tags.Clear()
                            updatedTask.Tags.AddRange(request.Task.Tags)
                        | "assigned_to" -> updatedTask.AssignedTo <- request.Task.AssignedTo
                        | "due_date" -> updatedTask.DueDate <- request.Task.DueDate
                        | _ -> () // Ignore unknown fields
                    )
                    
                    tasks.[updatedTask.Id] <- updatedTask
                    
                    logger.LogInformation("Updated task {TaskId}: {Title}", updatedTask.Id, updatedTask.Title)
                    return updatedTask
                }
                |> function
                | Ok task -> task
                | Error ex -> raise ex
        } |> Async.StartAsTask
    
    override this.DeleteTask(request: DeleteTaskRequest, context: ServerCallContext) =
        async {
            logger.LogInformation("DeleteTask called with ID: {TaskId}", request.Id)
            
            return
                result {
                    let! _ = validateTaskId request.Id
                    
                    match tasks.TryRemove(request.Id) with
                    | true, removedTask -> 
                        logger.LogInformation("Deleted task {TaskId}: {Title}", removedTask.Id, removedTask.Title)
                        return Empty()
                    | false, _ -> 
                        return! Error (RpcException(Status(StatusCode.NotFound, sprintf "Task with ID '%s' not found" request.Id)))
                }
                |> function
                | Ok empty -> empty
                | Error ex -> raise ex
        } |> Async.StartAsTask
    
    override this.WatchTasks(requestStream: IAsyncStreamReader<WatchTasksRequest>, responseStream: IServerStreamWriter<TaskEvent>, context: ServerCallContext) =
        async {
            logger.LogInformation("WatchTasks called")
            
            try
                // F# async enumeration over the request stream
                let rec processRequests () = async {
                    let! hasNext = requestStream.MoveNext() |> Async.AwaitTask
                    if hasNext && not context.CancellationToken.IsCancellationRequested then
                        let request = requestStream.Current
                        let now = Timestamp.FromDateTime(DateTime.UtcNow)
                        
                        // Determine which tasks to send based on request using pattern matching
                        let tasksToSend = 
                            match request.WatchAll, request.TaskIds.Count, String.IsNullOrEmpty(request.AssignedTo) with
                            | true, _, _ -> 
                                tasks.Values |> Seq.toList
                            | false, count, _ when count > 0 ->
                                request.TaskIds 
                                |> Seq.choose (fun taskId -> 
                                    match tasks.TryGetValue(taskId) with
                                    | true, task -> Some task
                                    | false, _ -> None)
                                |> Seq.toList
                            | false, 0, false ->
                                tasks.Values 
                                |> Seq.filter (fun t -> t.AssignedTo = request.AssignedTo)
                                |> Seq.toList
                            | _ -> []
                        
                        // Send events for selected tasks
                        for task in tasksToSend do
                            let taskEvent = TaskEvent(
                                EventType = EventType.Updated,
                                Task = task,
                                Timestamp = now
                            )
                            do! responseStream.WriteAsync(taskEvent) |> Async.AwaitTask
                        
                        // Continue processing
                        return! processRequests()
                }
                
                do! processRequests()
                
            with
            | ex -> 
                logger.LogError(ex, "Error in WatchTasks")
                raise (RpcException(Status(StatusCode.Internal, "Internal server error")))
        } |> Async.StartAsTask :> System.Threading.Tasks.Task

File: code/cpp/grpc/server/src/task_service.cpp
#include "task_service.h"
#include <algorithm>
#include <chrono>
#include <sstream>
#include <iomanip>

TaskServiceImpl::TaskServiceImpl() : next_id_(1) {
    // Create a few sample tasks
    auto now = std::chrono::system_clock::now();
    auto timestamp = std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count();
    
    // Sample task 1
    auto task1 = std::make_unique<tasks::Task>();
    task1->set_id("1");
    task1->set_title("Implement gRPC server");
    task1->set_description("Create a gRPC server for the task management API");
    task1->set_status(tasks::TaskStatus::IN_PROGRESS);
    task1->set_priority(tasks::TaskPriority::HIGH);
    task1->add_tags("grpc");
    task1->add_tags("server");
    task1->add_tags("cpp");
    task1->set_assigned_to("dev-team");
    task1->mutable_created_at()->set_seconds(timestamp - 3600);
    task1->mutable_updated_at()->set_seconds(timestamp - 1800);
    
    // Sample task 2
    auto task2 = std::make_unique<tasks::Task>();
    task2->set_id("2");
    task2->set_title("Write unit tests");
    task2->set_description("Add comprehensive unit tests for the task service");
    task2->set_status(tasks::TaskStatus::PENDING);
    task2->set_priority(tasks::TaskPriority::MEDIUM);
    task2->add_tags("testing");
    task2->add_tags("unit-tests");
    task2->set_assigned_to("qa-team");
    task2->mutable_created_at()->set_seconds(timestamp - 7200);
    task2->mutable_updated_at()->set_seconds(timestamp - 7200);
    
    tasks_["1"] = std::move(task1);
    tasks_["2"] = std::move(task2);
    next_id_ = 3;
}

grpc::Status TaskServiceImpl::ListTasks(grpc::ServerContext* context, 
                                       const tasks::ListTasksRequest* request,
                                       tasks::ListTasksResponse* response) {
    std::lock_guard<std::mutex> lock(tasks_mutex_);
    
    std::vector<const tasks::Task*> filtered_tasks;
    
    // Filter tasks based on request criteria
    for (const auto& [id, task] : tasks_) {
        if (matches_filter(*task, *request)) {
            filtered_tasks.push_back(task.get());
        }
    }
    
    // Sort tasks (simplified - by ID for now)
    std::sort(filtered_tasks.begin(), filtered_tasks.end(),
              [](const tasks::Task* a, const tasks::Task* b) {
                  return a->id() < b->id();
              });
    
    // Apply pagination
    int page_size = request->page_size() > 0 ? std::min(request->page_size(), 100) : 20;
    int start_idx = 0;
    
    if (!request->page_token().empty()) {
        try {
            start_idx = std::stoi(request->page_token());
        } catch (const std::exception&) {
            return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Invalid page token");
        }
    }
    
    int end_idx = std::min(start_idx + page_size, static_cast<int>(filtered_tasks.size()));
    
    // Add tasks to response
    for (int i = start_idx; i < end_idx; ++i) {
        *response->add_tasks() = *filtered_tasks[i];
    }
    
    response->set_total_count(filtered_tasks.size());
    
    // Set next page token if there are more results
    if (end_idx < static_cast<int>(filtered_tasks.size())) {
        response->set_next_page_token(std::to_string(end_idx));
    }
    
    return grpc::Status::OK;
}

grpc::Status TaskServiceImpl::GetTask(grpc::ServerContext* context, 
                                     const tasks::GetTaskRequest* request,
                                     tasks::Task* response) {
    std::lock_guard<std::mutex> lock(tasks_mutex_);
    
    auto it = tasks_.find(request->id());
    if (it == tasks_.end()) {
        return grpc::Status(grpc::StatusCode::NOT_FOUND, "Task not found");
    }
    
    *response = *it->second;
    return grpc::Status::OK;
}

grpc::Status TaskServiceImpl::CreateTask(grpc::ServerContext* context, 
                                        const tasks::CreateTaskRequest* request,
                                        tasks::Task* response) {
    if (request->title().empty()) {
        return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Title is required");
    }
    
    std::lock_guard<std::mutex> lock(tasks_mutex_);
    
    auto task = std::make_unique<tasks::Task>();
    task->set_id(generate_id());
    task->set_title(request->title());
    
    if (!request->description().empty()) {
        task->set_description(request->description());
    }
    
    task->set_status(tasks::TaskStatus::PENDING);
    
    if (request->has_priority()) {
        task->set_priority(request->priority());
    } else {
        task->set_priority(tasks::TaskPriority::MEDIUM);
    }
    
    // Copy tags
    for (const auto& tag : request->tags()) {
        task->add_tags(tag);
    }
    
    if (!request->assigned_to().empty()) {
        task->set_assigned_to(request->assigned_to());
    }
    
    // Set timestamps
    auto now = std::chrono::system_clock::now();
    auto timestamp = std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count();
    task->mutable_created_at()->set_seconds(timestamp);
    task->mutable_updated_at()->set_seconds(timestamp);
    
    std::string task_id = task->id();
    *response = *task;
    
    tasks_[task_id] = std::move(task);
    
    return grpc::Status::OK;
}

grpc::Status TaskServiceImpl::UpdateTask(grpc::ServerContext* context, 
                                        const tasks::UpdateTaskRequest* request,
                                        tasks::Task* response) {
    std::lock_guard<std::mutex> lock(tasks_mutex_);
    
    auto it = tasks_.find(request->id());
    if (it == tasks_.end()) {
        return grpc::Status(grpc::StatusCode::NOT_FOUND, "Task not found");
    }
    
    auto& task = *it->second;
    
    // Update fields based on field mask (simplified - update all provided fields)
    if (!request->title().empty()) {
        task.set_title(request->title());
    }
    
    if (!request->description().empty()) {
        task.set_description(request->description());
    }
    
    if (request->has_status()) {
        task.set_status(request->status());
    }
    
    if (request->has_priority()) {
        task.set_priority(request->priority());
    }
    
    if (!request->tags().empty()) {
        task.clear_tags();
        for (const auto& tag : request->tags()) {
            task.add_tags(tag);
        }
    }
    
    if (!request->assigned_to().empty()) {
        task.set_assigned_to(request->assigned_to());
    }
    
    // Update timestamp
    auto now = std::chrono::system_clock::now();
    auto timestamp = std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count();
    task.mutable_updated_at()->set_seconds(timestamp);
    
    *response = task;
    return grpc::Status::OK;
}

grpc::Status TaskServiceImpl::DeleteTask(grpc::ServerContext* context, 
                                        const tasks::DeleteTaskRequest* request,
                                        tasks::DeleteTaskResponse* response) {
    std::lock_guard<std::mutex> lock(tasks_mutex_);
    
    auto it = tasks_.find(request->id());
    if (it == tasks_.end()) {
        return grpc::Status(grpc::StatusCode::NOT_FOUND, "Task not found");
    }
    
    tasks_.erase(it);
    response->set_success(true);
    
    return grpc::Status::OK;
}

grpc::Status TaskServiceImpl::UpdateTaskStatus(grpc::ServerContext* context, 
                                              const tasks::UpdateTaskStatusRequest* request,
                                              tasks::Task* response) {
    std::lock_guard<std::mutex> lock(tasks_mutex_);
    
    auto it = tasks_.find(request->id());
    if (it == tasks_.end()) {
        return grpc::Status(grpc::StatusCode::NOT_FOUND, "Task not found");
    }
    
    auto& task = *it->second;
    task.set_status(request->status());
    
    // Update timestamp
    auto now = std::chrono::system_clock::now();
    auto timestamp = std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count();
    task.mutable_updated_at()->set_seconds(timestamp);
    
    *response = task;
    return grpc::Status::OK;
}

std::string TaskServiceImpl::generate_id() {
    return std::to_string(next_id_++);
}

std::unique_ptr<tasks::Task> TaskServiceImpl::create_task_copy(const tasks::Task& task) {
    auto copy = std::make_unique<tasks::Task>();
    *copy = task;
    return copy;
}

bool TaskServiceImpl::matches_filter(const tasks::Task& task, const tasks::ListTasksRequest& request) {
    // Filter by status
    if (request.has_status() && task.status() != request.status()) {
        return false;
    }
    
    // Filter by assigned_to
    if (!request.assigned_to().empty() && task.assigned_to() != request.assigned_to()) {
        return false;
    }
    
    // Filter by tags (task must have all requested tags)
    for (const auto& requested_tag : request.tags()) {
        bool tag_found = false;
        for (const auto& task_tag : task.tags()) {
            if (task_tag == requested_tag) {
                tag_found = true;
                break;
            }
        }
        if (!tag_found) {
            return false;
        }
    }
    
    return true;
}